spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala

```diff
@@ -0,0 +1,160 @@
+        t.canDeleteUsingMetadataWhere(dataSourceFilters)
+      case _ => false
+    }
+  }
+
+  private def isMetadataOnlyCondition(table: Table, cond: Expression): Boolean = {
+    val partitionColumnRefs = table.partitioning.flatMap(_.references)
+    val partitionColumnNames = partitionColumnRefs.map(_.fieldNames.mkString(".")).toSet
+
+    val resolver = SQLConf.get.resolver
+    val predicates = splitConjunctivePredicates(cond)
+
+    predicates.forall { pred =>
+      val isExecutable = pred.deterministic && !SubqueryExpression.hasSubquery(pred)
+      isExecutable && isMetadataPredicate(pred, partitionColumnNames, resolver)
+    }
+  }
+
+  private def isMetadataPredicate(
+      predicate: Expression,
+      partitionColumnNames: Set[String],
+      resolver: Resolver): Boolean = {
+    predicate.references.forall { reference => partitionColumnNames.exists(resolver(reference.name, _)) }
+  }
+
+  private def newWriteInfo(schema: StructType): LogicalWriteInfo = {
+    val uuid = UUID.randomUUID()
+    LogicalWriteInfoImpl(queryId = uuid.toString, schema, CaseInsensitiveStringMap.empty)
+  }
+
+  private def pushDownFilters(
+      builder: RowLevelOperationsBuilder,
+      cond: Expression,
+      output: Seq[AttributeReference]): Unit = {
+
+    builder match {
+      case s: SupportsPushDownRowLevelFilters =>
+        val dataSourceFilters = toDataSourceFilters(cond, output)
+        s.pushFilters(dataSourceFilters)
+      case _ => // do nothing
+    }
+  }
+
+  private def toDataSourceFilters(
+      cond: Expression,
+      output: Seq[AttributeReference]): Array[sources.Filter] = {
+
+    val predicates = splitConjunctivePredicates(cond)
+    val normalizedPredicates = DataSourceStrategy.normalizeExprs(predicates, output)
+    toDataSourceFilters(normalizedPredicates)
+  }
+
+  private def toDataSourceFilters(
+      predicates: Seq[Expression],
+      skipUntranslated: Boolean = true): Array[sources.Filter] = {
+
+    predicates.flatMap { p =>
+      val translatedFilter = DataSourceStrategy.translateFilter(p, supportNestedPredicatePushdown = true)
+      if (translatedFilter.isEmpty && !skipUntranslated) {
+        throw new AnalysisException(s"Cannot translate expression to source filter: $p")
+      }
+      translatedFilter
+    }.toArray
+  }
+
+  private def buildFilterPlan(cond: Expression, scanRelation: DataSourceV2ScanRelation): LogicalPlan = {
+    val fileNameExpr = Alias(InputFileName(), FILE_NAME_COL)()
+    val fileNameProjection = Project(scanRelation.output :+ fileNameExpr, scanRelation)
+    val matchingFilter = Filter(cond, fileNameProjection)
+    val fileAttr = findOutputAttr(matchingFilter, FILE_NAME_COL)
+    val matchingFileAgg = Aggregate(Seq(fileAttr), Seq(fileAttr), matchingFilter)
+    // TODO: inject DynamicFileFilter only if either Scan or Write support it
+    val dynamicFileFilter = DynamicFileFilter(scanRelation, matchingFileAgg)
```

Comment:
You might want to separate this into two clear sections: the delete plan with the `notMatchingRowFilter` and the dynamic filter plan with `matchingFilter` and `matchingFileAgg`. I was expecting the former when I started reading this method.